src/lib.rs

```diff
@@ -88,54 +87,35 @@
 //!
 //! The previous example triggers a fail point by modifying the `FAILPOINT`
 //! environment variable. In practice, you'll often want to trigger fail points
-//! programmatically, in unit tests. Unfortunately, unit testing with fail
-//! points is complicated by concurrency concerns, so requires some careful
-//! setup. Fail points are global resources, and Rust tests run in parallel,
+//! programmatically, in unit tests.
+//! Fail points are global resources, and Rust tests run in parallel,
 //! so tests that exercise fail points generally need to hold a lock to
-//! avoid interfering with each other.
+//! avoid interfering with each other. This is accomplished by `FailScenario`.
 //!
 //! Here's a basic pattern for writing unit tests tests with fail points:
 //!
-//! ```
-//! #[macro_use]
-//! extern crate lazy_static;
-//! #[macro_use]
-//! extern crate fail;
-//!
-//! use std::sync::{Mutex, MutexGuard};
+//! ```rust
+//! use fail::{fail_point, FailScenario};
 //!
 //! fn do_fallible_work() {
 //!     fail_point!("read-dir");
 //!     let _dir: Vec<_> = std::fs::read_dir(".").unwrap().collect();
 //!     // ... do some work on the directory ...
 //! }
 //!
-//! lazy_static! {
-//!     static ref LOCK: Mutex<()> = Mutex::new(());
-//! }
-//!
-//! fn setup<'a>() -> MutexGuard<'a, ()> {
-//!     let guard = LOCK.lock().unwrap_or_else(|e| e.into_inner());
-//!     fail::teardown();
-//!     fail::setup();
-//!     guard
-//! }
-//!
 //! #[test]
 //! #[should_panic]
 //! fn test_fallible_work() {
-//!     let _gaurd = setup();
-//!     fail::cfg("read-dir", "panic").unwrap();
+//!     let scenario = FailScenario::setup();
+//!     scenario.cfg("read-dir", "panic").unwrap();
```

Comment:
It looks to me like in this patch `cfg` is still a free function. Can you make it a method on `FailScenario` as depicted here?

Reply:
That was my initial approach, but I had to revert back to a free function. This doc error is a leftover I didn't catch when undoing.

The reason why this needs to be a free function is that there is a case for consumers like this one in tikv: https://github.com/tikv/tikv/blob/82de1403905cb12ed988d85fa8da2ab99fb0adde/src/server/service/debug.rs#L269.
There, `fail::cfg` is used in project logic to override fail-points configuration, if any. That is decoupled from the scenario, thus we can't use `FailScenario::cfg` there.

Reply:
If both `fail::cfg` and `fail_point!` can be made as methods of `FailScenario`, maybe we don't need the global lock at all. And tests can be run in parallel.

But I'm happy with current approach as it makes life easier at least.

Reply:
Maybe this could be another PR? :)

Reply:
Yes, of course.

Changing them into methods also loses the flexibility of macros, which can be put to anywhere you want without any context. However, it maybe worthy to speedup tests when failpoint are used heavily like TiKV.

Reply:
I think this is the same topic as https://github.com/pingcap/fail-rs/issues/24. I guess, we can brainstorm that design and its implication there a bit more. From what I can see, it will be a major/invasive change anyway.